hypothesis-python/src/hypothesis/provisional.py

```diff
@@ -20,62 +20,162 @@
+
+URL_SAFE_CHARACTERS = frozenset(string.ascii_letters + string.digits + "$-_.+!*'(),")
+
+
+# This file is sourced from http://data.iana.org/TLD/tlds-alpha-by-domain.txt
+# The file contains additional information about the date that it was last updated.
+with open(
+    os.path.join(os.path.dirname(__file__), "vendor", "tlds-alpha-by-domain.txt")
+) as tld_file:
+    __header = next(tld_file)
+    assert __header.startswith("#")
+    TOP_LEVEL_DOMAINS = sorted((line.rstrip() for line in tld_file), key=len)
+TOP_LEVEL_DOMAINS.insert(0, "COM")
+
+
+class DomainNameStrategy(SearchStrategy):
+    @staticmethod
+    def clean_inputs(minimum, maximum, value, variable_name):
+        if value is None:
+            value = maximum
+        elif not isinstance(value, int):
+            raise InvalidArgument(
+                "Expected integer but %s is a %s"
+                % (variable_name, type(value).__name__)
+            )
+        elif not minimum <= value <= maximum:
+            raise InvalidArgument(
+                "Invalid value %r < %s=%r < %r"
+                % (minimum, variable_name, value, maximum)
+            )
+        return value
+
+    def __init__(self, max_length=None, max_element_length=None):
+        """
+        A strategy for :rfc:`1035` fully qualified domain names.
+
+        The upper limit for max_length is 255 in accordance with :rfc:`1035#section-2.3.4`
+        The lower limit for max_length is 4, corresponding to a two letter domain
+        with a single letter sub-domain.
+        The upper limit for max_element_length is 63 in accordance with :rfc:`1035#section-2.3.4`
+        The lower limit for max_element_length is 1 in accordance with :rfc:`1035#section-2.3.4`
+        """
+        # https://tools.ietf.org/html/rfc1035#section-2.3.4
+
+        max_length = self.clean_inputs(4, 255, max_length, "max_length")
+        max_element_length = self.clean_inputs(
+            1, 63, max_element_length, "max_element_length"
+        )
+
+        super(DomainNameStrategy, self).__init__()
+        self.max_length = max_length
+        self.max_element_length = max_element_length
+
+        # These regular expressions are constructed to match the documented
+        # information in https://tools.ietf.org/html/rfc1035#section-2.3.1
+        # which defines the allowed syntax of a sub-domain string.
+        if self.max_element_length == 1:
+            self.label_regex = r"[a-zA-Z]"
+        elif self.max_element_length == 2:
+            self.label_regex = r"[a-zA-Z][a-zA-Z0-9]?"
+        else:
+            max_center_chars = self.max_element_length - 2
+            self.label_regex = (
+                r"[a-zA-Z]([a-zA-Z0-9\-]{0,%s}[a-zA-Z0-9])?" % max_center_chars
+            )
+
+    def do_draw(self, data):
+        # 1 - Select a valid top-level domain (TLD) name
+        # 2 - Check that the number of characters in our selected TLD wont
+        # prevent us from generating at least a 1 character sub-domain.
```

Comment:
nit:

```suggestion
        # prevent us from generating at least a 1 character subdomain.
```

You use "subdomain" without a hyphen in the next comment!

Reply:
Ive switched everything to `subdomain`